### ISSUE TITLE
Endless loop when ChannelOutboundBuffer.addFlush

### ISSUE THREAD

**Issue by allenzhu** [netty/netty#3938]
Hi, on Netty 4.0.25, i face a endless loop when `ChannelOutboundBuffer.addFlush`:

```
"NettyServerNIOSelectorForPiled_3_3" prio=10 tid=0x00007f11b0043000 nid=0x1e62a runnable [0x000000004b947000]
   java.lang.Thread.State: RUNNABLE
        at io.netty.channel.ChannelOutboundBuffer.addFlush(ChannelOutboundBuffer.java:149)
        at io.netty.channel.AbstractChannel$AbstractUnsafe.flush(AbstractChannel.java:684)
        at io.netty.channel.DefaultChannelPipeline$HeadContext.flush(DefaultChannelPipeline.java:1118)
        at io.netty.channel.AbstractChannelHandlerContext.invokeFlush(AbstractChannelHandlerContext.java:663)
        at io.netty.channel.AbstractChannelHandlerContext.access$1500(AbstractChannelHandlerContext.java:32)
        at io.netty.channel.AbstractChannelHandlerContext$16.run(AbstractChannelHandlerContext.java:651)
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:380)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:357)
        at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:116)
        at java.lang.Thread.run(Thread.java:662)
```

And cpu one core utility ratio 100%:

```
$top -H -p 119263

top - 18:16:24 up 247 days,  6:35,  1 user,  load average: 1.11, 1.05, 1.01
Tasks: 308 total,   2 running, 306 sleeping,   0 stopped,   0 zombie
Cpu(s):  4.4%us,  0.4%sy,  0.0%ni, 95.1%id,  0.1%wa,  0.0%hi,  0.0%si,  0.0%st
Mem:  99164684k total, 98508940k used,   655744k free,   242208k buffers
Swap:        0k total,        0k used,        0k free, 86805380k cached

   PID USER      PR  NI  VIRT  RES  SHR S %CPU %MEM    TIME+  COMMAND                                                                                
124458 admin     20   0  497g  86g  81g R 100.2 91.1   2405:39 java                                                                                  
119309 admin     20   0  497g  86g  81g R  0.7 91.1  15:48.39 java                                                                                   
120128 admin     20   0  497g  86g  81g S  0.3 91.1   5:16.30 java              
```

This cause the other task attach to this io thread can not execute at all, even client close the channel.

client side net state(Send-Q full):

```
tcp        0 130321 10.97.240.138:6661          10.97.240.137:10911         FIN_WAIT1   
tcp        0 130321 10.97.240.138:31856         10.97.240.137:10911         FIN_WAIT1   
```

server side net state(Recv-Q full):

```
tcp   235152      0 10.97.240.137:10911         10.97.240.138:31856         ESTABLISHED 
tcp   236112      0 10.97.240.137:10911         10.97.240.138:6661          ESTABLISHED 
```

Attach the process found that entry.next(id=2593) == itself(id=2593), also flushedEntry==unflushedEntry==tailEntry, so this cause the endless loop.

```
this    ChannelOutboundBuffer  (id=2592)    
    channel NioSocketChannel  (id=2595) 
    fireChannelWritabilityChangedTask   null    
    flushed 808346277   
    flushedEntry    ChannelOutboundBuffer$Entry  (id=2593)  
    inFail  false   
    nioBufferCount  0   
    nioBufferSize   0   
    tailEntry   ChannelOutboundBuffer$Entry  (id=2593)  
    totalPendingSize    132 
    unflushedEntry  ChannelOutboundBuffer$Entry  (id=2593)  
    unwritable  0   

entry   ChannelOutboundBuffer$Entry  (id=2593)  
    buf null    
    bufs    null    
    cancelled   false   
    count   -1  
    handle  Recycler$DefaultHandle  (id=2607)   
    msg ManyMessageTransfer  (id=2608)  
    next    ChannelOutboundBuffer$Entry  (id=2593)  
    pendingSize 0   
    progress    0   
    promise DefaultChannelPromise  (id=2625)    
    total   2075
```

By the way, app migrate a channel from one EventLoopGroup to another before this situation. Code as below:

```
    @Sharable
    class NettyServerHandler extends SimpleChannelInboundHandler<RemotingCommand> {

        @Override
        protected void channelRead0(ChannelHandlerContext ctx, RemotingCommand msg) throws Exception {
            Attribute<Integer> attr = ctx.channel().attr(key);
            Integer status = attr.get();
            if (null != status && MARK_PILED == status) {
                boolean pResult = attr.compareAndSet(status, SET_PILED);
                if (pResult) {
                    long start = System.currentTimeMillis();
                    ChannelFuture dFuture = ctx.deregister();
                    dFuture.addListener(new ChannelFutureListener() {
                        @Override
                        public void operationComplete(ChannelFuture future) throws Exception {
                            log.warn("deregister channel complete, channel={}, thread={}, pipeline={}",
                                future.channel(), Thread.currentThread(), future.channel().pipeline());

                            ChannelFuture rf = eventLoopGroupSelectorForPiled.register(future.channel());
                            rf.addListener(new ChannelFutureListener() {
                                @Override
                                public void operationComplete(ChannelFuture rFuture) throws Exception {
                                    log.warn("register channel complete, channel={}, thread={}, pipeline={}",
                                        rFuture.channel(), Thread.currentThread(), rFuture.channel()
                                            .pipeline());
                                }
                            });
                        }
                    });
                    long cost = System.currentTimeMillis() - start;
                    log.warn("start migrate channel, channel={}, pipeline={}, thread={}, cost={}",
                        ctx.channel(), ctx.channel().pipeline(), Thread.currentThread(), cost);
                }
                else {
                    log.warn("start migrate channel fail because of status change, channel={}, thread={}",
                        ctx.channel(), Thread.currentThread());
                }
            }
            processMessageReceived(ctx, msg);
        }
    }
```

May there are something wrong with the usage or something else ?

**Comment by normanmaurer** [comment]
Will investigate. Can you reproduce this easily? Also vould you upgrade to 4.0.29.Final and check if you still see the problem?

**Comment by allenzhu** [comment]
Yes, i can reproduce it. but it need run for a while and also i will try 4.0.29.Final

**Comment by normanmaurer** [comment]
If you could provide a simple reproducer it would help a lot

**Comment by allenzhu** [comment]
reproduce step are:
1. start broker.
2. start five consumer to pull message (high throughput), after a while this happen.(The main change code as shown previously. The other part can see this: https://github.com/alibaba/RocketMQ/tree/master/rocketmq-remoting/src/main/java/com/alibaba/rocketmq/remoting)

**Comment by ThomasLau** [comment]
wow,it's strange here,i think the only chance that tailEntry.next=itself is when it is assinged to a value:
```java
public void addMessage(Object msg, int size, ChannelPromise promise) {
        Entry entry = Entry.newInstance(msg, size, total(msg), promise);
        ...
        Entry tail = tailEntry;
        tail.next = entry;
        tailEntry = entry;
```
entry.next is null here.i guess this endless loop may be caused by another exception.

**Comment by allenzhu** [comment]
yeah, it looks like that, or maybe concurrently operate the data structure somewhere but i didn't find it.
And i try 4.0.29.Final version to run the application for more than 15 hours, it doesn't reproduce util now.

**Comment by normanmaurer** [comment]
Keep us posted when you can reproduce it .

**Comment by allenzhu** [comment]
Ok, by the way the usage of migrating channel from one EventLoopGroup to another is suitable?
Are there some traps about it ?

**Comment by allenzhu** [comment]
@normanmaurer i reproduce it with 4.0.29.Final. 
The same phenomenon as before. One cpu core util ratio 100%.

**Comment by normanmaurer** [comment]
@allenzhu and it only happens if you move from one EventLoop to another ?

**Comment by allenzhu** [comment]
Yeah, @normanmaurer i found a problem that will broke the rule of serial execute per one Channel when moving from one EventLoop to another.
childHandler init as follows:

    .childHandler(new ChannelInitializer<SocketChannel>() {
    	@Override
    	public void initChannel(SocketChannel ch) throws Exception {
    	    ch.pipeline().addLast(
		defaultEventExecutorGroup, //
		new NettyEncoder(), //
		new NettyDecoder(), //
		new IdleStateHandler(0, 0, nettyServerConfig.getServerChannelMaxIdleTimeSeconds()),//
		new NettyConnetManageHandler(), //
		new NettyServerHandler());
	}

The migrating logic in NettyServerHandler shown in early comment.

When a request comes, the original NioEventLoop will read the data(```NioByteUnsafe.read```) and ```pipeline.fireChannelRead(byteBuf)```, after that triger ```pipeline.fireChannelReadComplete()```. When fireChannelRead will execute ```NettyServerHandler.channelRead0``` to triger migrating.
There will be two NioEventLoop attach to one NioChannel, Deregister task will be add to original EventLoop taskQueue, Register task will be add to new EventLoop taskQueue. 

```ctx.deregister()``` --> ```HeadContext.deregister``` --> ```AbstractNioChannel.doDeregister``` cancel SelectionKey.

    protected void doDeregister() throws Exception {
        eventLoop().cancel(selectionKey());
    }

```eventLoopGroupSelectorForPiled.register``` --> ```MultithreadEventLoopGroup.register``` --> ```AbstractUnsafe.register0``` --> ```AbstractNioChannel.doRegister```.
In ```AbstractNioChannel.doRegister``` register ops with 0.

    selectionKey = javaChannel().register(eventLoop().selector, 0, this);

```AbstractNioChannel.doBeginRead``` can register ```SelectionKey.OP_READ```, let's look at ```AbstractUnsafe.register0```:

    private void register0(ChannelPromise promise) {
        try {
    	// check if the channel is still open as it could be closed in the mean time when the register
    	// call was outside of the eventLoop
    	if (!promise.setUncancellable() || !ensureOpen(promise)) {
    	    return;
    	}
    	boolean firstRegistration = neverRegistered;
    	doRegister();
    	neverRegistered = false;
    	registered = true;
    	safeSetSuccess(promise);
    	pipeline.fireChannelRegistered();
    	// Only fire a channelActive if the channel has never been registered. This prevents firing
    	// multiple channel actives if the channel is deregistered and re-registered.
    	if (firstRegistration && isActive()) {
    	    pipeline.fireChannelActive();
    	}
    ....

```pipeline.fireChannelActive()``` can fire ```AbstractNioChannel.doBeginRead``` invoke, but the channel is new created so firstRegistration=false.

There is a chance to do so in ```pipeline.fireChannelReadComplete()``` --> ```AbstractNioChannel.doBeginRead```.  Let's see the code:

    protected void doBeginRead() throws Exception {
        // Channel.read() or ChannelHandlerContext.read() was called
        if (inputShutdown) {
            return;
        }

        final SelectionKey selectionKey = this.selectionKey;
        if (!selectionKey.isValid()) {
            return;
        }

        readPending = true;

        final int interestOps = selectionKey.interestOps();
        if ((interestOps & readInterestOp) == 0) {
            selectionKey.interestOps(interestOps | readInterestOp);
        }
    }

so if not exec ```doRegister()``` ,the selectionKey will be  invalid  and will return without register SelectionKey.OP_READ.  ```pipeline.fireChannelReadComplete()``` and ```doRegister()``` will be invoked in diffrent thread, so there is a race condition. This will cause new EventLoop can not read data from this channel, because it do not registe READ ops. 

The endless loop mentioned early may be the same reason. 

Look at ```AbstractNioByteChannel.doWrite``` code:

	....
	if (done) {
	    in.remove();
	} else {
	    incompleteWrite(setOpWrite);
	    break;
	}
	....

if write incomplete, it will schedule flush again later. so this may cause flush in old eventLoop, ```outboundBuffer.addMessage``` in new eventLoop at the same time. Concurrently operate outboundBuffer and get unexpected result. To prove this, i add some log in ```addMessage``` and reproduce:

    public void addMessage(Object msg, int size, ChannelPromise promise) {
        Entry entry = Entry.newInstance(msg, size, total(msg), promise);
        if (entry.next != null) {
            System.out.println("never expected 2, entry=" + entry + ", entry.next=" + entry.next);
        }
        Entry tmpTail = tailEntry;
        if (tailEntry == null) {
            flushedEntry = null;
            tailEntry = entry;
        } else {
            Entry tail = tailEntry;
            tail.next = entry;
            tailEntry = entry;
            if (entry == tail || flushedEntry == unflushedEntry) {
                StringBuilder sb = new StringBuilder("never expected 1, entry=");
                sb.append(entry).append(", tail=").append(tail).append(", tmpTail=").append(tmpTail);
                sb.append(", flushedEntry=").append(flushedEntry).append(", unflushedEntry=").append(unflushedEntry);
                sb.append(", flushed=").append(flushed).append(", channel=").append(channel);
                sb.append(", msg=").append(msg).append(", outbound=").append(this).append(", date=");
                sb.append(new java.util.Date()).append(", thread=").append(Thread.currentThread());
                System.out.println(sb.toString());
            }
        }
        ....

Result entry==tail, entry.next=itself:

    never expected 1, entry=io.netty.channel.ChannelOutboundBuffer$Entry@15cab027, tail=io.netty.channel.ChannelOutboundBuffer$Entry@15cab027, tmpTail=io.netty.channel.ChannelOutboundBuffer$Entry@15cab027, flushedEntry=null, unflushedEntry=null, flushed=3, channel=[id: 0xc6910e1c, /10.101.88.105:35426 => /10.97.220.25:10911], msg=com.alibaba.rocketmq.broker.pagecache.ManyMessageTransfer@2df2db9e, outbound=io.netty.channel.ChannelOutboundBuffer@233079a3, date=Thu Jul 09 03:52:58 CST 2015, thread=Thread[NettyServerNIOSelectorForPiled_6_5,5,main]

Is there such a problem?

**Comment by allenzhu** [comment]
@normanmaurer, is this a problem or my usage is wrong ?

**Comment by normanmaurer** [comment]
@allenzhu sorry but I had no time yet to have a deeper look.

**Comment by allenzhu** [comment]
@normanmaurer fine, take a look when you are free :)

**Comment by allenzhu** [comment]
@normanmaurer , is there such a problem ?

**Comment by normanmaurer** [comment]
@allenzhu sorry still had no time... but from what I read here there may be some race possible.

**Comment by allenzhu** [comment]
@normanmaurer yeah, it looks like that

**Comment by johnou** [comment]
@allenzhu you might want to try 4.0.35.Final when it is released. @normanmaurer end of month? sooner?

**Comment by normanmaurer** [comment]
This was fixed in 5.0 duets removal of moving Channels between EventLoops.
